crates/uv/src/commands/project/lock.rs

```diff
@@ -180,59 +179,133 @@ pub(super) async fn do_lock(
+    let lock = read_lockfile(workspace, upgrade).await?;
+    let LockedRequirements { preferences, git } = lock
+        .as_ref()
+        .map(|lock| read_lock_requirements(lock, upgrade))
+        .unwrap_or_default();
 
     // Create the Git resolver.
     let git = GitResolver::from_refs(git);
 
-    // Create a build dispatch.
-    let build_dispatch = BuildDispatch::new(
-        &client,
-        cache,
-        interpreter,
-        index_locations,
-        &flat_index,
-        &index,
-        &git,
-        &in_flight,
-        index_strategy,
-        setup_py,
-        config_setting,
-        build_isolation,
-        link_mode,
-        build_options,
-        exclude_newer,
-        concurrency,
-        preview,
-    );
-
-    // Resolve the requirements.
-    let resolution = pip::operations::resolve(
-        requirements,
-        constraints,
-        overrides,
-        dev,
-        source_trees,
-        None,
-        &extras,
-        preferences,
-        EmptyInstalledPackages,
-        &hasher,
-        &Reinstall::default(),
-        upgrade,
-        None,
-        None,
-        python_requirement,
-        &client,
-        &flat_index,
-        &index,
-        &build_dispatch,
-        concurrency,
-        options,
-        printer,
-        preview,
-    )
-    .await?;
+    // Try resolving from the lockfile.
+    let resolution = match lock {
+        None => None,
+        Some(lock) => {
+            // Prefill the index with the lockfile.
+            let index = lock.to_index(workspace)?;
+
+            // Create a build dispatch.
+            //
+            // When resolving from the lockfile we can still download/install new distributions,
+            // but we rely on the lockfile for the metadata of any existing distributions. If we
+            // have any outdated metadata we fall back to a clean resolve.
+            let build_dispatch = BuildDispatch::new(
+                &client,
+                cache,
+                interpreter,
+                index_locations,
+                &flat_index,
+                &index,
+                &git,
+                &in_flight,
+                index_strategy,
+                setup_py,
+                config_setting,
+                build_isolation,
+                link_mode,
+                build_options,
+                exclude_newer,
+                concurrency,
+                preview,
+            );
+
+            // TODO(ibraheem): avoid cloning the manifest here
+            pip::operations::resolve(
+                requirements.clone(),
+                constraints.clone(),
+                overrides.clone(),
+                dev.clone(),
+                source_trees.clone(),
+                None,
+                &extras,
+                preferences.clone(),
+                EmptyInstalledPackages,
+                &hasher,
+                &Reinstall::default(),
+                upgrade,
```

Comment:
We probably need to do some rigorous thinking around how `reinstall` and `upgrade` work here, and whether they will be respected correctly, both for the registry distributions that are already locked and for the mutable distributions that we're omitting (like path dependencies).

Reply:
I think this is worth some dedicated tests or at least some exploration on the command-line.